zerver/tests/test_tusd.py

```diff
@@ -0,0 +1,185 @@
+            body,
+            content_type="application/json",
+            HTTP_HOOK_NAME="pre-create",
+        )
+        self.assert_json_error_contains(
+            result, "Not logged in: API authentication or user session required", 401
+        )
+
+    def test_tusd_pre_create_hook(self) -> None:
+        self.login("hamlet")
+        body = {
+            "Upload": {
+                "ID": "",
+                "IsFinal": False,
+                "IsPartial": False,
+                "MetaData": {
+                    "filename": "jupyterlab-galata-report.zip",
+                    "filetype": "application/zip",
+                    "name": "jupyterlab-galata-report.zip",
+                    "type": "application/zip",
+                },
+                "Offset": 0,
+                "PartialUploads": None,
+                "Size": settings.MAX_FILE_UPLOAD_SIZE * 1024 * 1024 - 100,
+                "SizeIsDeferred": False,
+            },
+        }
+        result = self.client_post(
+            f"/tusd/hooks?secret={settings.SHARED_SECRET}",
+            body,
+            content_type="application/json",
+            HTTP_HOOK_NAME="pre-create",
+        )
+        self.assert_json_success(result)
+
+    def test_file_too_big_failure(self) -> None:
+        self.login("hamlet")
+        body = {
+            "Upload": {
+                "Size": settings.MAX_FILE_UPLOAD_SIZE * 1024 * 1024 + 100,
+            },
+        }
+        with self.settings(MAX_FILE_UPLOAD_SIZE=0):
+            result = self.client_post(
+                f"/tusd/hooks?secret={settings.SHARED_SECRET}",
+                body,
+                content_type="application/json",
+                HTTP_HOOK_NAME="pre-create",
+            )
+
+        self.assert_json_error(result, "Uploaded file is larger than the allowed limit of 0 MiB")
+
+    def test_tusd_pre_finish_hook(self) -> None:
+        self.login("hamlet")
+        assert settings.MAX_FILE_UPLOAD_SIZE is not None
+        assert settings.LOCAL_FILES_DIR is not None
+        assert settings.LOCAL_UPLOADS_DIR is not None
+        file_id = "e4b4acc5ddb8675d3af4e75a1a095bd7"
+        file_path = os.path.join(settings.LOCAL_UPLOADS_DIR, "tusd", file_id)
+        os.makedirs(os.path.dirname(file_path), exist_ok=True)
+        with open(file_path, "+bw") as f:
+            f.write(os.urandom(100))
+            f.close()
+        self.assertTrue(os.path.exists(file_path))
+        body = {
+            "Upload": {
+                "ID": file_id,
+                "IsFinal": False,
+                "IsPartial": False,
+                "MetaData": {
+                    "filename": "brijsiyag.zip",
+                    "filetype": "application/zip",
+                    "name": "brijsiyag.zip",
+                    "type": "application/zip",
+                },
+                "Offset": 0,
+                "PartialUploads": None,
+                "Size": settings.MAX_FILE_UPLOAD_SIZE * 1024 * 1024 - 100,
```

Comment:
I'm not sure it's helpful to lie here, when we know the size is 100 bytes.

Reply:
updated.